src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns_Constant.cs

```diff
@@ -0,0 +1,159 @@
+                        long value = constantValue.Int64Value;
+                        return (value < int.MinValue) ? AsLargeNegativeHelper(node) :
+                               (value > int.MaxValue) ? AsLargePositiveHelper(node) :  AsIntValueHelper(node);
+                    }
+                case ConstantValueTypeDiscriminator.UInt32:
+                case ConstantValueTypeDiscriminator.UInt64:
+                    {
+                        // this can be inside the range of ints or outside the range at the positive side.
+                        ulong value = constantValue.UInt64Value;
+                        return (value > int.MaxValue) ? AsLargePositiveHelper(node) : AsIntValueHelper(node);
+                    }
+                default:
+                    {
+                        // all of the remaining integral types have value sets totally contained in the values of int.
+                        return AsIntValueHelper(node);
+                    }
+            }
+        }
+
+        private MethodSymbol AsLargeNegativeHelper(CSharpSyntaxNode node)
+        {
+            return MakePrivateImplementationHelper(
+                node,
+                PrivateImplementationDetails.AsLargeNegativeName,
+                (m, pi) => new SynthesizedAsLargeNegativeMethod(m, pi));
+        }
+
+        private MethodSymbol AsLargePositiveHelper(CSharpSyntaxNode node)
+        {
+            return MakePrivateImplementationHelper(
+                node,
+                PrivateImplementationDetails.AsLargePositiveName,
+                (m, pi) => new SynthesizedAsLargePositiveMethod(m, pi));
+        }
+
+        private MethodSymbol AsIntValueHelper(CSharpSyntaxNode node)
+        {
+            return MakePrivateImplementationHelper(
+                node,
+                PrivateImplementationDetails.AsIntValueName,
+                (m, pi) => new SynthesizedAsIntValueMethod(m, pi));
+        }
+
+        private MethodSymbol MakePrivateImplementationHelper(
+            CSharpSyntaxNode node,
+            string helperName,
+            Func<Emit.PEModuleBuilder, PrivateImplementationDetails, SynthesizedGlobalMethodSymbol> makeHelper)
+        {
+            var module = this.EmitModule;
+            if (module == null)
+            {
+                return null;
+            }
+
+            // If we have already generated the helper, possibly for another switch
+            // or on another thread, we don't need to regenerate it.
+            var privateImplClass = module.GetPrivateImplClass(node, _diagnostics);
+            var helper = (MethodSymbol)privateImplClass.GetMethod(helperName);
+            if (helper != (object)null)
+            {
+                return helper;
+            }
+
+            helper = makeHelper(module, privateImplClass);
+            privateImplClass.TryAddSynthesizedMethod(helper);
+            return helper;
```

Comment:
This doesn't look thread-safe, other thread could have added a different symbol and we are continuing using the one that nobody tracks.